[ConstantInt] Disable implicit truncation in ConstantInt::get()

nikic opened the pull request:
Disable implicit truncation in the ConstantInt constructor by default. This means that it needs to be passed a signed/unsigned (depending on the IsSigned flag) value matching the bit width.

The intention is to prevent the recurring bug where people write something like `ConstantInt::get(Ty, -1)`, and this "works" until `Ty` is larger than 64-bit and then the value is incorrect due to missing type extension.

This is the continuation of https://github.com/llvm/llvm-project/pull/112670, which originally allowed implicit truncation in this constructor to reduce initial scope of the change.

nikic committed:
[IR] Disable implicit truncation in ConstantInt::get() by default

RKSimon reviewed (commented):
SGTM - just a couple of queries.

What about ConstantInt::getSigned - have you considered adding a ImplicitTrunc flag to that as well?

Also, we've only got a few weeks before next release branch (including the quiet holiday period with less reports), and its likely random assert regression bugs will inevitably appear for a few months - are we OK with that (and merging fixes etc.)?

nikic commented:
> Also, we've only got a few weeks before next release branch (including the quiet holiday period with less reports), and its likely random assert regression bugs will inevitably appear for a few months - are we OK with that (and merging fixes etc.)?

My plan was to land this change on main and then revert it on the release branch.

nikic commented:
> What about ConstantInt::getSigned - have you considered adding a ImplicitTrunc flag to that as well?

Done in https://github.com/llvm/llvm-project/pull/172875.

nikic commented:
Next crash, also in LSR...
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define i32 @main(i16 %.pre.pre.i, i16 %inc18.i.i.peel) {
entry:
  %spec.select.i.peel = add i16 %.pre.pre.i, %inc18.i.i.peel
  br label %lbl_1468.i

lbl_1468.i:                                       ; preds = %lbl_1468.i, %entry
  %.pre.i = phi i16 [ %spec.select.i.peel, %entry ], [ %inc18.i.i.i, %lbl_1468.i ]
  %inc18.i.i.i = add i16 %.pre.i, 1
  %tobool.not1281.i.i.i.i = icmp eq i16 %inc18.i.i.i, 0
  br i1 false, label %func_117.exit.loopexit, label %lbl_1468.i

func_117.exit.loopexit:                           ; preds = %lbl_1468.i
  store i16 %.pre.i, ptr null, align 2
  ret i32 0
}
```

nikic commented:
Added the test case in https://github.com/llvm/llvm-project/commit/0809a9927d67d9484e19cf3d88541bd79c5ba3b4 and adjusted the remaining `ConstantInt::get()` use in LSR...

dtcxzyw reviewed (approved):
I think my existing fuzzing corpus cannot discover more violations. It should be fine to land and wait for crash reports from other continuous fuzzing systems.
Remember to notify the release manager to revert this patch in the next release.

RKSimon reviewed (approved):
LGTM (with a suitable merge conflict fix) - no objections to either immediate committal (+ reversion in 22.x) or waiting for 22.x to branch

topperc reviewed (approved):
LGTM

mikaelholmen commented:
> @mikaelholmen Thanks, should be fixed by [faa7ede](https://github.com/llvm/llvm-project/commit/faa7ede811d345eeee5aa66ad1cdb38b51222a16).

Yep, thanks!

nathanchance commented:
I see an assertion failure when building the Linux kernel for PowerPC after this change.

`cvise` spits out:

```c
typedef unsigned u32;
enum { false };
enum radeon_family {
  CHIP_RV710,
  CHIP_CEDAR,
  CHIP_JUNIPER,
  CHIP_SUMO,
  CHIP_SUMO2
};
struct evergreen_asic {
  unsigned tile_config;
};
struct radeon_device {
  struct evergreen_asic config;
  enum radeon_family family;
} r100_mm_wreg(struct radeon_device *, u32, u32, _Bool),
    evergreen_gpu_init_rdev;
u32 evergreen_gpu_init_sq_config, evergreen_gpu_init_mc_arb_ramcfg,
    evergreen_gpu_init_sq_gpr_resource_mgmt_3, evergreen_gpu_init_smx_dc_ctl0,
    evergreen_gpu_init_vgt_cache_invalidation,
    evergreen_gpu_init_sq_lds_resource_mgmt,
    evergreen_gpu_init_sq_stack_resource_mgmt_3,
    evergreen_gpu_init_gb_addr_config, evergreen_gpu_init_disabled_rb_mask,
    evergreen_gpu_init_sq_stack_resource_mgmt_1, evergreen_gpu_init_sx_debug_1,
    evergreen_gpu_init_hdp_host_path_cntl,
    evergreen_gpu_init_sq_gpr_resource_mgmt_1,
    evergreen_gpu_init_sq_stack_resource_mgmt_2,
    evergreen_gpu_init_sq_gpr_resource_mgmt_2,
    evergreen_gpu_init_sq_thread_resource_mgmt, evergreen_gpu_init_tmp,
    evergreen_gpu_init_sq_thread_resource_mgmt_2;
int evergreen_gpu_init_i, evergreen_gpu_init_ps_thread_count,
    evergreen_gpu_init_j;
void evergreen_gpu_init() {
  switch (evergreen_gpu_init_rdev.family) {
  case CHIP_JUNIPER:
    evergreen_gpu_init_gb_addr_config = 2;
  default:;
  }
  for (evergreen_gpu_init_i = evergreen_gpu_init_j = 0; evergreen_gpu_init_i;
       evergreen_gpu_init_i += 8)
    evergreen_gpu_init_mc_arb_ramcfg = 0;
  {
    switch (evergreen_gpu_init_mc_arb_ramcfg) {}
  }
  evergreen_gpu_init_rdev.config.tile_config |=
      evergreen_gpu_init_gb_addr_config;
  evergreen_gpu_init_tmp = 0;
  evergreen_gpu_init_disabled_rb_mask = evergreen_gpu_init_tmp;
  if (evergreen_gpu_init_disabled_rb_mask)
    ;
  evergreen_gpu_init_sx_debug_1 = 0;
  r100_mm_wreg(&evergreen_gpu_init_rdev, 0, evergreen_gpu_init_sx_debug_1,
               false);
  evergreen_gpu_init_smx_dc_ctl0 = 0;
  r100_mm_wreg(&evergreen_gpu_init_rdev, 0, evergreen_gpu_init_smx_dc_ctl0,
               false);
  evergreen_gpu_init_sq_config = evergreen_gpu_init_sq_lds_resource_mgmt =
      evergreen_gpu_init_sq_gpr_resource_mgmt_1 =
          evergreen_gpu_init_sq_gpr_resource_mgmt_2 =
              evergreen_gpu_init_sq_gpr_resource_mgmt_3 =
                  evergreen_gpu_init_ps_thread_count =
                      evergreen_gpu_init_sq_thread_resource_mgmt =
                          evergreen_gpu_init_sq_thread_resource_mgmt_2 =
                              evergreen_gpu_init_sq_stack_resource_mgmt_1 =
                                  evergreen_gpu_init_sq_stack_resource_mgmt_2 =
                                      evergreen_gpu_init_sq_stack_resource_mgmt_3 =
                                          0;
  r100_mm_wreg(&evergreen_gpu_init_rdev, 0, evergreen_gpu_init_sq_config,
               false);
  switch (evergreen_gpu_init_rdev.family) {
  case CHIP_CEDAR:
  case CHIP_SUMO:
  case CHIP_SUMO2:
    evergreen_gpu_init_vgt_cache_invalidation = 1;
    break;
  default:
    evergreen_gpu_init_vgt_cache_invalidation = 0;
  }
  r100_mm_wreg(&evergreen_gpu_init_rdev, 0,
               evergreen_gpu_init_vgt_cache_invalidation, false);
  evergreen_gpu_init_hdp_host_path_cntl = 0;
  r100_mm_wreg(&evergreen_gpu_init_rdev, 0,
               evergreen_gpu_init_hdp_host_path_cntl, false);
}

```

```
$ clang --target=powerpc-linux-gnu -O2 -c -o /dev/null evergreen.i
clang: llvm/include/llvm/ADT/APInt.h:128: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isUIntN(BitWidth, val) && "Value is not an N-bit unsigned value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang --target=powerpc-linux-gnu -O2 -c -o /dev/null evergreen.i
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'evergreen.i'.
4.	Running pass 'Split GEPs to a variadic base and a constant offset for better CSE' on function '@evergreen_gpu_init'
 #0 0x0000557f1edb7448 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (clang-22+0x3969448)
 #1 0x0000557f1edb4bd5 llvm::sys::RunSignalHandlers() (clang-22+0x3966bd5)
 #2 0x0000557f1ed34f26 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007fa46aa3e4d0 (/usr/lib/libc.so.6+0x3e4d0)
 #4 0x00007fa46aa9890c (/usr/lib/libc.so.6+0x9890c)
 #5 0x00007fa46aa3e3a0 raise (/usr/lib/libc.so.6+0x3e3a0)
 #6 0x00007fa46aa2557a abort (/usr/lib/libc.so.6+0x2557a)
 #7 0x00007fa46aa254e3 __assert_perror_fail (/usr/lib/libc.so.6+0x254e3)
 #8 0x0000557f1db12a4e llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) X86ISelLowering.cpp:0:0
 #9 0x0000557f1e79c5e2 llvm::ConstantInt::get(llvm::Type*, unsigned long, bool, bool) (clang-22+0x334e5e2)
#10 0x0000557f1ecbfae3 (anonymous namespace)::SeparateConstOffsetFromGEP::run(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#11 0x0000557f1ecc1c53 (anonymous namespace)::SeparateConstOffsetFromGEPLegacyPass::runOnFunction(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#12 0x0000557f1e8893d8 llvm::FPPassManager::runOnFunction(llvm::Function&) (clang-22+0x343b3d8)
#13 0x0000557f1e890af2 llvm::FPPassManager::runOnModule(llvm::Module&) (clang-22+0x3442af2)
#14 0x0000557f1e889dc0 llvm::legacy::PassManagerImpl::run(llvm::Module&) (clang-22+0x343bdc0)
#15 0x0000557f1f539330 clang::emitBackendOutput(clang::CompilerInstance&, clang::CodeGenOptions&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (clang-22+0x40eb330)
#16 0x0000557f1f54e598 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (clang-22+0x4100598)
#17 0x0000557f20ba8d19 clang::ParseAST(clang::Sema&, bool, bool) (clang-22+0x575ad19)
#18 0x0000557f1fa521d6 clang::FrontendAction::Execute() (clang-22+0x46041d6)
#19 0x0000557f1f9b92fd clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (clang-22+0x456b2fd)
#20 0x0000557f1fb2d492 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (clang-22+0x46df492)
#21 0x0000557f1d8f9eca cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (clang-22+0x24abeca)
#22 0x0000557f1d8f5ab5 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) driver.cpp:0:0
#23 0x0000557f1d8f808d int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::$_0>(long, llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#24 0x0000557f1f81ac49 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#25 0x0000557f1ed34c0e llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (clang-22+0x38e6c0e)
#26 0x0000557f1f81a483 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (clang-22+0x43cc483)
#27 0x0000557f1f7db0ac clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (clang-22+0x438d0ac)
#28 0x0000557f1f7db2c7 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (clang-22+0x438d2c7)
#29 0x0000557f1f7f82e8 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (clang-22+0x43aa2e8)
#30 0x0000557f1d8f5370 clang_main(int, char**, llvm::ToolContext const&) (clang-22+0x24a7370)
#31 0x0000557f1d9059c7 main (clang-22+0x24b79c7)
#32 0x00007fa46aa27635 (/usr/lib/libc.so.6+0x27635)
#33 0x00007fa46aa276e9 __libc_start_main (/usr/lib/libc.so.6+0x276e9)
#34 0x0000557f1d8f3525 _start (clang-22+0x24a5525)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
ClangBuiltLinux clang version 22.0.0git (https://github.com/llvm/llvm-project.git a83c89495ba6fe0134dcaa02372c320cc7ff0dbf)
...
```

I further reduced that C reproducer with `llvm-reduce` down to:

```llvm
target datalayout = "E-m:e-p:32:32-Fn32-i64:64-n32"
target triple = "powerpc-unknown-linux-gnu"

define ptr @evergreen_gpu_init() {
entry:
  %switch.tableidx = add i32 0, -1
  %switch.gep = getelementptr i32, ptr null, i32 %switch.tableidx
  ret ptr %switch.gep
}

```

```
$ llc -o /dev/null reduced.ll
llc: llvm/include/llvm/ADT/APInt.h:128: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isUIntN(BitWidth, val) && "Value is not an N-bit unsigned value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
Stack dump:
0.	Program arguments: llc -o /dev/null reduced.ll
1.	Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.	Running pass 'Split GEPs to a variadic base and a constant offset for better CSE' on function '@evergreen_gpu_init'
 #0 0x0000562d5a4424e8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (llc+0x36a04e8)
 #1 0x0000562d5a43fc05 llvm::sys::RunSignalHandlers() (llc+0x369dc05)
 #2 0x0000562d5a443251 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007f7f5f63e4d0 (/usr/lib/libc.so.6+0x3e4d0)
 #4 0x00007f7f5f69890c (/usr/lib/libc.so.6+0x9890c)
 #5 0x00007f7f5f63e3a0 raise (/usr/lib/libc.so.6+0x3e3a0)
 #6 0x00007f7f5f62557a abort (/usr/lib/libc.so.6+0x2557a)
 #7 0x00007f7f5f6254e3 __assert_perror_fail (/usr/lib/libc.so.6+0x254e3)
 #8 0x0000562d58b18e2e llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) X86ISelLowering.cpp:0:0
 #9 0x0000562d59888772 llvm::ConstantInt::get(llvm::Type*, unsigned long, bool, bool) (llc+0x2ae6772)
#10 0x0000562d59fb54f3 (anonymous namespace)::SeparateConstOffsetFromGEP::run(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#11 0x0000562d59fb7663 (anonymous namespace)::SeparateConstOffsetFromGEPLegacyPass::runOnFunction(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#12 0x0000562d59999ee8 llvm::FPPassManager::runOnFunction(llvm::Function&) (llc+0x2bf7ee8)
#13 0x0000562d599a1812 llvm::FPPassManager::runOnModule(llvm::Module&) (llc+0x2bff812)
#14 0x0000562d5999a960 llvm::legacy::PassManagerImpl::run(llvm::Module&) (llc+0x2bf8960)
#15 0x0000562d588f1dd5 compileModule(char**, llvm::SmallVectorImpl<llvm::PassPlugin>&, llvm::LLVMContext&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>&) llc.cpp:0:0
#16 0x0000562d588ef303 main (llc+0x1b4d303)
#17 0x00007f7f5f627635 (/usr/lib/libc.so.6+0x27635)
#18 0x00007f7f5f6276e9 __libc_start_main (/usr/lib/libc.so.6+0x276e9)
#19 0x0000562d588eab25 _start (llc+0x1b48b25)

```

```
# bad: [fb7e80570d0beb2f6eb252e53aa0ba62d5c83ab0] [clang-doc] Add protected members to class template (#174883)
# good: [af8bb1d5219353a099d63f9da6c9dd5dff77c600] Fix Bazel build for 4dc9a0e (#174691)
git bisect start 'fb7e80570d0beb2f6eb252e53aa0ba62d5c83ab0' 'af8bb1d5219353a099d63f9da6c9dd5dff77c600'
# bad: [94d4e3ae1c65f5eee9f017a787908767ddcb5ef7] [libc++] Refactor poll_with_backoff (#173184)
git bisect bad 94d4e3ae1c65f5eee9f017a787908767ddcb5ef7
# bad: [17e6f51e51d36c0f784f6c400680e4aebac12d13] [GitHub] Add more glob patterns for `mlir:python` label (#174711)
git bisect bad 17e6f51e51d36c0f784f6c400680e4aebac12d13
# bad: [a83c89495ba6fe0134dcaa02372c320cc7ff0dbf] Reapply [ConstantInt] Disable implicit truncation in ConstantInt::get() (#171456)
git bisect bad a83c89495ba6fe0134dcaa02372c320cc7ff0dbf
# good: [050195053c8aff21bb02ade4c90cb7f42343849c] Revert "[UniformityAnalysis] Remove an incorrect assertion in uniformity analysis (#174117)"
git bisect good 050195053c8aff21bb02ade4c90cb7f42343849c
# good: [eb13822b511f59399cbb09301a6c1f25b2be0850] [AMDGPU] Modifies cvt and atomic builtin def to take _Float16 for HIP/C++ (#174707)
git bisect good eb13822b511f59399cbb09301a6c1f25b2be0850
# good: [c85b8ff4d74bea08ac690a6c2a612313241237cc] [MemRef] Add dim reification for AssumeAlignmentOp (#174477)
git bisect good c85b8ff4d74bea08ac690a6c2a612313241237cc
# good: [164cfda88e7a339d0c4beed6bb0a5fe3c25c73a2] [BOLT][AArch64] Add rseq test (#174413)
git bisect good 164cfda88e7a339d0c4beed6bb0a5fe3c25c73a2
# first bad commit: [a83c89495ba6fe0134dcaa02372c320cc7ff0dbf] Reapply [ConstantInt] Disable implicit truncation in ConstantInt::get() (#171456)
```

nathanchance commented:
> @nathanchance Thanks for the report, should be fixed by [1c07ba6](https://github.com/llvm/llvm-project/commit/1c07ba612d03b05b2ff72f36bd5da6b732f931e4).

Thanks, this cleared up the majority of the instances I saw with that configuration. One more from the same pass it seems, even with 1c07ba612d03b05b2ff72f36bd5da6b732f931e4 applied.

```c
void *ecc_swap_digits_in;
void get_unaligned_be64(void *);
void ecc_swap_digits() {
  long *src = ecc_swap_digits_in;
  int i;
  for (i = 0;; i++)
    get_unaligned_be64(&src[-1 - i]);
}
```

[The original C function](https://elixir.bootlin.com/linux/v6.18.4/source/include/crypto/internal/ecc.h#L63) had a parameter and `ndigits - 1 - i` as the array index.

```
$ clang --target=powerpc-linux-gnu -O2 -c -o /dev/null ecc.i
clang: llvm/include/llvm/ADT/APInt.h:121: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isIntN(BitWidth, val) && "Value is not an N-bit signed value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang --target=powerpc-linux-gnu -O2 -c -o /dev/null ecc.i
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'ecc.i'.
4.	Running pass 'Split GEPs to a variadic base and a constant offset for better CSE' on function '@ecc_swap_digits'
 #0 0x00005573ba782188 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (clang-22+0x396c188)
 #1 0x00005573ba77f915 llvm::sys::RunSignalHandlers() (clang-22+0x3969915)
 #2 0x00005573ba6ff556 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007fe80323e4d0 (/usr/lib/libc.so.6+0x3e4d0)
 #4 0x00007fe80329890c (/usr/lib/libc.so.6+0x9890c)
 #5 0x00007fe80323e3a0 raise (/usr/lib/libc.so.6+0x3e3a0)
 #6 0x00007fe80322557a abort (/usr/lib/libc.so.6+0x2557a)
 #7 0x00007fe8032254e3 __assert_perror_fail (/usr/lib/libc.so.6+0x254e3)
 #8 0x00005573b94dc5f3 llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) X86ISelLowering.cpp:0:0
 #9 0x00005573ba165c52 llvm::ConstantInt::get(llvm::Type*, unsigned long, bool, bool) (clang-22+0x334fc52)
#10 0x00005573ba68a106 (anonymous namespace)::SeparateConstOffsetFromGEP::run(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#11 0x00005573ba68c283 (anonymous namespace)::SeparateConstOffsetFromGEPLegacyPass::runOnFunction(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#12 0x00005573ba252a18 llvm::FPPassManager::runOnFunction(llvm::Function&) (clang-22+0x343ca18)
#13 0x00005573ba25a132 llvm::FPPassManager::runOnModule(llvm::Module&) (clang-22+0x3444132)
#14 0x00005573ba253400 llvm::legacy::PassManagerImpl::run(llvm::Module&) (clang-22+0x343d400)
#15 0x00005573baf05660 clang::emitBackendOutput(clang::CompilerInstance&, clang::CodeGenOptions&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (clang-22+0x40ef660)
#16 0x00005573baf1a8d8 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (clang-22+0x41048d8)
#17 0x00005573bc5780c9 clang::ParseAST(clang::Sema&, bool, bool) (clang-22+0x57620c9)
#18 0x00005573bb41f536 clang::FrontendAction::Execute() (clang-22+0x4609536)
#19 0x00005573bb3865cd clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (clang-22+0x45705cd)
#20 0x00005573bb4fa7d2 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (clang-22+0x46e47d2)
#21 0x00005573b92c3a8a cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (clang-22+0x24ada8a)
#22 0x00005573b92bf675 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) driver.cpp:0:0
#23 0x00005573b92c1c4d int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::$_0>(long, llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#24 0x00005573bb1e7809 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#25 0x00005573ba6ff23e llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (clang-22+0x38e923e)
#26 0x00005573bb1e7043 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (clang-22+0x43d1043)
#27 0x00005573bb1a7c6c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (clang-22+0x4391c6c)
#28 0x00005573bb1a7e87 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (clang-22+0x4391e87)
#29 0x00005573bb1c4ea8 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (clang-22+0x43aeea8)
#30 0x00005573b92bef30 clang_main(int, char**, llvm::ToolContext const&) (clang-22+0x24a8f30)
#31 0x00005573b92cf5b7 main (clang-22+0x24b95b7)
#32 0x00007fe803227635 (/usr/lib/libc.so.6+0x27635)
#33 0x00007fe8032276e9 __libc_start_main (/usr/lib/libc.so.6+0x276e9)
#34 0x00005573b92bd0e5 _start (clang-22+0x24a70e5)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
ClangBuiltLinux clang version 22.0.0git (https://github.com/llvm/llvm-project.git 1c07ba612d03b05b2ff72f36bd5da6b732f931e4)
```

LLVM IR:

```llvm
target datalayout = "E-m:e-p:32:32-Fn32-i64:64-n32"
target triple = "powerpc-unknown-linux-gnu"

define void @ecc_swap_digits() {
entry:
  br label %for.cond

for.cond:                                         ; preds = %for.cond, %entry
  %i.0 = phi i32 [ 0, %entry ], [ %inc, %for.cond ]
  %sub = xor i32 %i.0, -1
  %arrayidx = getelementptr i32, ptr null, i32 %sub
  %inc = add nsw i32 %i.0, 1
  br label %for.cond
}
```

```
llc: llvm/include/llvm/ADT/APInt.h:121: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isIntN(BitWidth, val) && "Value is not an N-bit signed value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
Stack dump:
0.	Program arguments: llc -o /dev/null reduced.ll
1.	Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.	Running pass 'Split GEPs to a variadic base and a constant offset for better CSE' on function '@ecc_swap_digits'
 #0 0x000055bd284ac338 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (llc+0x36a2338)
 #1 0x000055bd284a9a55 llvm::sys::RunSignalHandlers() (llc+0x369fa55)
 #2 0x000055bd284ad0a1 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x00007fd15be3e4d0 (/usr/lib/libc.so.6+0x3e4d0)
 #4 0x00007fd15be9890c (/usr/lib/libc.so.6+0x9890c)
 #5 0x00007fd15be3e3a0 raise (/usr/lib/libc.so.6+0x3e3a0)
 #6 0x00007fd15be2557a abort (/usr/lib/libc.so.6+0x2557a)
 #7 0x00007fd15be254e3 __assert_perror_fail (/usr/lib/libc.so.6+0x254e3)
 #8 0x000055bd26b82763 llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) X86ISelLowering.cpp:0:0
 #9 0x000055bd278f1dc2 llvm::ConstantInt::get(llvm::Type*, unsigned long, bool, bool) (llc+0x2ae7dc2)
#10 0x000055bd2801e906 (anonymous namespace)::SeparateConstOffsetFromGEP::run(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#11 0x000055bd28020a83 (anonymous namespace)::SeparateConstOffsetFromGEPLegacyPass::runOnFunction(llvm::Function&) SeparateConstOffsetFromGEP.cpp:0:0
#12 0x000055bd27a03508 llvm::FPPassManager::runOnFunction(llvm::Function&) (llc+0x2bf9508)
#13 0x000055bd27a0ae32 llvm::FPPassManager::runOnModule(llvm::Module&) (llc+0x2c00e32)
#14 0x000055bd27a03f80 llvm::legacy::PassManagerImpl::run(llvm::Module&) (llc+0x2bf9f80)
#15 0x000055bd2695b755 compileModule(char**, llvm::SmallVectorImpl<llvm::PassPlugin>&, llvm::LLVMContext&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>&) llc.cpp:0:0
#16 0x000055bd26958c83 main (llc+0x1b4ec83)
#17 0x00007fd15be27635 (/usr/lib/libc.so.6+0x27635)
#18 0x00007fd15be276e9 __libc_start_main (/usr/lib/libc.so.6+0x276e9)
#19 0x000055bd269544a5 _start (llc+0x1b4a4a5)
```

nikic commented:
@gulfemsavrun Thanks, should be fixed by https://github.com/llvm/llvm-project/commit/2fe39f24b2899cc71aabfad1dc8558f0754b5480.

nathanchance commented:
@nikic One more that only appeared after testing `ARCH=hexagon`:

```c
enum { FORMAT_STATE_NUM };
struct {
} vsnprintf_spec;
unsigned state;
struct {
} vsnprintf_fmt;
char vsnprintf_str, vsnprintf_end, vsnprintf_old_fmt;
long vsnprintf_num;
int vsnprintf(char *buf, unsigned size, const char *fmt_str,
              __builtin_va_list args) {
  switch (vsnprintf_fmt, state) {
    int read = vsnprintf_old_fmt;
    read;
  case FORMAT_STATE_NUM:
    __builtin_va_arg(args, long long);
    vsnprintf_num, vsnprintf_end, vsnprintf_spec;
    goto out;
  }
out:
  return &vsnprintf_str - buf;
}
```

```
$ clang --target=hexagon-linux-musl -O2 -Wno-unused-value -S -emit-llvm -Xclang -disable-llvm-passes vsprintf.i
clang: llvm/include/llvm/ADT/APInt.h:128: llvm::APInt::APInt(unsigned int, uint64_t, bool, bool): Assertion `llvm::isUIntN(BitWidth, val) && "Value is not an N-bit unsigned value"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang --target=hexagon-linux-musl -O2 -Wno-unused-value -S -emit-llvm -Xclang -disable-llvm-passes vsprintf.i
1.	<eof> parser at end of file
2.	vsprintf.i:9:5: LLVM IR generation of declaration 'vsnprintf'
3.	vsprintf.i:9:5: Generating code for declaration 'vsnprintf'
4.	vsprintf.i:11:33: LLVM IR generation of compound statement ('{}')
 #0 0x00005643d2c46ba8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (clang-23+0x6069ba8)
 #1 0x00005643d2c44355 llvm::sys::RunSignalHandlers() (clang-23+0x6067355)
 #2 0x00005643d2bc3136 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007f7ed703e4d0 (/usr/lib/libc.so.6+0x3e4d0)
 #4 0x00007f7ed709890c (/usr/lib/libc.so.6+0x9890c)
 #5 0x00007f7ed703e3a0 raise (/usr/lib/libc.so.6+0x3e3a0)
 #6 0x00007f7ed702557a abort (/usr/lib/libc.so.6+0x2557a)
 #7 0x00007f7ed70254e3 __assert_perror_fail (/usr/lib/libc.so.6+0x254e3)
 #8 0x00005643d0ad537e llvm::APInt::APInt(unsigned int, unsigned long, bool, bool) AArch64ISelLowering.cpp:0:0
 #9 0x00005643d2629045 llvm::ConstantInt::get(llvm::IntegerType*, unsigned long, bool, bool) (clang-23+0x5a4c045)
#10 0x00005643d364a489 (anonymous namespace)::HexagonABIInfo::EmitVAArg(clang::CodeGen::CodeGenFunction&, clang::CodeGen::Address, clang::QualType, clang::CodeGen::AggValueSlot) const Hexagon.cpp:0:0
#11 0x00005643d2f1c505 clang::CodeGen::CodeGenFunction::EmitVAArg(clang::VAArgExpr*, clang::CodeGen::Address&, clang::CodeGen::AggValueSlot) (clang-23+0x633f505)
#12 0x00005643d30927aa (anonymous namespace)::ScalarExprEmitter::VisitVAArgExpr(clang::VAArgExpr*) CGExprScalar.cpp:0:0
#13 0x00005643d308489d clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (clang-23+0x64a789d)
#14 0x00005643d304cfed clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) (clang-23+0x646ffed)
#15 0x00005643d2fe31d2 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x64061d2)
#16 0x00005643d2ff0fc6 clang::CodeGen::CodeGenFunction::EmitCaseStmt(clang::CaseStmt const&, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x6413fc6)
#17 0x00005643d2fe3bfb clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x6406bfb)
#18 0x00005643d2fe30df clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x64060df)
#19 0x00005643d2fe3b81 clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x6406b81)
#20 0x00005643d2fe30df clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x64060df)
#21 0x00005643d2fe8234 clang::CodeGen::CodeGenFunction::EmitSwitchStmt(clang::SwitchStmt const&) (clang-23+0x640b234)
#22 0x00005643d2fe3591 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (clang-23+0x6406591)
#23 0x00005643d2ff1171 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (clang-23+0x6414171)
#24 0x00005643d2fcef9c clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (clang-23+0x63f1f9c)
#25 0x00005643d2e97764 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (clang-23+0x62ba764)
#26 0x00005643d2e8ef3f clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (clang-23+0x62b1f3f)
#27 0x00005643d2e940da clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (clang-23+0x62b70da)
#28 0x00005643d2e8da31 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (clang-23+0x62b0a31)
#29 0x00005643d33f21cc (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) ModuleBuilder.cpp:0:0
#30 0x00005643d33e9a76 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (clang-23+0x680ca76)
#31 0x00005643d4a5ffba clang::ParseAST(clang::Sema&, bool, bool) (clang-23+0x7e82fba)
#32 0x00005643d38deef6 clang::FrontendAction::Execute() (clang-23+0x6d01ef6)
#33 0x00005643d384581d clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (clang-23+0x6c6881d)
#34 0x00005643d39ba062 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (clang-23+0x6ddd062)
#35 0x00005643d0a715f4 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (clang-23+0x3e945f4)
#36 0x00005643d0a6d195 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) driver.cpp:0:0
#37 0x00005643d0a6f76d int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::$_0>(long, llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#38 0x00005643d36b6ee9 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#39 0x00005643d2bc2e1e llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (clang-23+0x5fe5e1e)
#40 0x00005643d36b6723 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (clang-23+0x6ad9723)
#41 0x00005643d367754c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (clang-23+0x6a9a54c)
#42 0x00005643d3677767 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (clang-23+0x6a9a767)
#43 0x00005643d36947d8 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (clang-23+0x6ab77d8)
#44 0x00005643d0a6c9e0 clang_main(int, char**, llvm::ToolContext const&) (clang-23+0x3e8f9e0)
#45 0x00005643d0a7d197 main (clang-23+0x3ea0197)
#46 0x00007f7ed7027635 (/usr/lib/libc.so.6+0x27635)
#47 0x00007f7ed70276e9 __libc_start_main (/usr/lib/libc.so.6+0x276e9)
#48 0x00005643d0a6aba5 _start (clang-23+0x3e8dba5)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
ClangBuiltLinux clang version 23.0.0git (https://github.com/llvm/llvm-project.git 8a8d65a0c59ab4d32eaded05fda0ce2ffdd25c39)
...
```
